# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ознакомимся с данными. Проверим, какие значения содержатся в столбце с информацией о задолженности. Если 0 и 1, значит,данные 
#заполнены корректно. 

data['debt'].unique()

array([0, 1])

In [23]:
data_children = data.pivot_table(index=['children'], values=['debt'], aggfunc=['sum', 'count'])
data_children['percent_debt'] = round(data_children['sum'] / data_children['count'] * 100, 2)
data_children

sum  count percent_debt
          debt   debt             
children                          
0         1063  14091         7.54
1          444   4808         9.23
2          194   2052         9.45
3           27    330         8.18
4            4     41         9.76
5            0      9         0.00

**Вывод:** 


Проведенное исследование показало, что среди заемщиков не имеющих детей и имеющих от одного до четырех детей, наименьшее количество должников в категории без детей. При этом, в категории с количеством детей равным 5, должников - 0%. Но нужно учесть, что в этой категории имеется всего 9 записей, при общем размере анализируемых данных более двадцати тысяч. 

По выборкам по с количеством детей от 1 до 4 мы имеем следующую картину:
- У заёмщиков с двумя детьми процент должников на 0,22% больше, чем у тех, кто имеет одного ребенка
- У заемщиков с 9,76% должников, что на 0,31% больше, чем у тех, кто имеет двоих детей
- Категория заёмщиков с тремя детьми нарушает эту последовательность, т.к. процент должников в этой категории существенно ниже, чем в других категориях с детьми. 

В результате можем сделать вывод, что отчасти зависимость между количеством детей и сроком возврата кредита прослеживается, если рассматривать категории, по которым имеем наибольшее количество данных (от 0 до 2 детей). Для получения же более точных выводов необходимо большее количество наблюдений по клиентам из категорий с 3, 4 и 5 детьми.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [31]:
data_status = data.pivot_table(index=['family_status'], values=['debt'], aggfunc=['sum', 'count'])
data_status['percent_debt'] = round(data_status['sum'] / data_status['count'] * 100, 2)
data_status

sum  count percent_debt
                      debt   debt             
family_status                                 
Не женат / не замужем  273   2796         9.76
в разводе               84   1189         7.06
вдовец / вдова          63    951         6.62
гражданский брак       385   4134         9.31
женат / замужем        927  12261         7.56

**Вывод:** 

В данном иссследовании мы делили заемщиков на группы по семейному статусу. Из полученных данных можно сделать следующие выводы:

Заемщики, которые никогда не вступали в брак (категории: "не женат / не замужем" и "гражданский брак"), иемют процент должников существенно выше, чем заемщики, которые в настоящее время или ранее состояли в официальном браке (категории: "женат / замужем", "вдовец / вдова", "в разводе").

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [30]:
data_income = data.pivot_table(index=['total_income_category'], values=['debt'], aggfunc=['sum', 'count'])
data_income['percent_debt'] = round(data_income['sum'] / data_income['count'] * 100, 2)
data_income

sum  count percent_debt
                       debt   debt             
total_income_category                          
A                         2     25         8.00
B                       354   5014         7.06
C                      1353  15921         8.50
D                        21    349         6.02
E                         2     22         9.09

**Вывод:** 

Из полученных данных видно, что процент должников в категории заемщиков с наименьшим доходом (кат. E) превышает остальные категории. Но в категории А процент должников также достаточно велик, хотя это категория заемщиков с самым большим доходом, превышающим 1 млн. руб. в месяц. Между категориями C и B соотношение должников распределилось с разницей примерно в 1,5% в пользу категории B. В целом, из полученных данных мы не можем сделать вывод о прямой зависимости между уровнем дохода и возвратом кредита в срок. Но, нужно учесть, что выборки в категориях A, D и E в сотни раз меньше, чем в категориях B и C. 
Вполне возможно, что при анализе сопоставимых по размеру выборок, мы получим другие данные и сможем увидеть зависимость количества должников от уровня дохода.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [32]:
data_purpose = data.pivot_table(index=['purpose_category'], values=['debt'], aggfunc=['sum', 'count'])
data_purpose['percent_debt'] = round(data_purpose['sum'] / data_purpose['count'] * 100, 2)
data_purpose

sum  count percent_debt
                         debt   debt             
purpose_category                                 
операции с автомобилем    400   4279         9.35
операции с недвижимостью  780  10751         7.26
получение образования     369   3988         9.25
проведение свадьбы        183   2313         7.91

**Вывод:** 
Исследование показало, что процент должников в категориях "операции с автомобилем" и "получение образования" чуть выше, чем в оставшихся двух категориях. Разница составляет: 1,5-2%.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*

Пропуски были обнаружены в столбцах, содержащих общий трудовой стаж в днях (days_employed) и ежемесячный доход (total_income). Количество пропусков в этих столбцах равно и составляет 2174. Можно предположить, что в датасет занесены только те данные, которые подтверждены справкой 2-НДФЛ, а сотрудники, по которым есть пропуски, такую справку не предоставили. Другой возможный вариант - сбой в работе системы, в результате которого не подтянулись данные из другой системы. Чтобы определить точную причину, нужно выяснить, каким образом заполняются данные по доходам и стажу.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Пропуски в количественных переменных сильно влияют на результат анализа данных и приводят в неверной интерпретации. Чтобы этого не допустить принято заполнять пропуски средним арифметическим или медианным значением. В исследуемом нами наборе данных пропуски были выявлены в столбцах, содежащих количество дней трудового стажа и размер ежемесячного дохода. В этих столбцах есть данные, которые сильно выделяются среди большинства. В таких случаях среднее значение даёт некорректную характеристику данных, поэтому для заполнения пропусков мы применили медианное значение.

### Шаг 4: общий вывод.

В данном проекте было проведено исследование надежности заемщиков банка. Входные данные от банка - статистика о платежеспособности клиентов. Результаты иссследования будут учтены при построении модели кредитного скоринга - — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Целью исследования является проверка нескольких гипотез:

- Гипотеза 1: количество детей влияет на факт погашения кредита в срок;
- Гипотеза 2: семейное положение влияет на факт погашения кредита в срок;
- Гипотеза 3: уровень дохода влияет на факт погашения кредита в срок;
- Гипотеза 4: цель кредита влияет на факт погашения кредита в срок;

В результате проверки гипотез мы установили:

**Гипотеза 1 частично подтверждена:**

- 7,54% доля просроченных кредитов у заемщиков без детей;
- 9,23% доля просроченных кредитов у заемщиков с одним ребенком;
- 9,45% доля просроченных кредитов у заемщиков с двумя детьми.

Для более точных выводов необходимо больше данных по клиентам с 3, 4 и 5 детьми.

Рекомендация: провести дополнительное исследование на большей выборке данных.

**Гипотеза 2 подтверждена:**

- Заемщики из категорий "не женат/не замужем" (9,76%) и "гражданский брак" (9,31%) иемют процент должников существенно выше, чем заемщики, которые в настоящее время или ранее состояли в официальном браке, из категорий "женат/замужем" (7,56%), "вдовец/вдова" (6,62%), "в разводе"(7,06%).

**Гипотеза 3 частично подтверждена:**

7,06% - доля просроченных кредитов у клиентов с уровнем дохода от 200.001 до 1.000.000
8,50% - доля просроченных кредитов у клиентов с уровнем дохода от до 50.001 до 200.000

Для более точных выводов необходимо больше данных по клиентам с уровнем дохода категорий A (до 30.000), E(от 1.000.001) и D(от 30.001 до 50.000).

Рекомендация: провести дополнительное исследование на большей выборке данных.

**Гипотеза 4 подтверждена:**

9,35% - доля просроченных кредитов, целью которых были операции с автомобилями
9,25% - доля просроченных кредитов, целью которых было получение образования
7,91% - доля просроченных кредитов, целью которых было проведение свадьбы
7,26% - доля просроченных кредитов, целью которых были операции с недвижимостью

**Общие рекомендации:**

1. Для построения модели кредитного скоринга можно использовать зависимости из второй и четвертой гипотез (семейное положение и цели кредита)

2. Необходимо собрать более полную выборку для повторной проверки первой и третьей гипотез.
